File: medium/0337_house_robber_iii/solution.py
from typing import Optional, Tuple

from utils.python.nodes import TreeNode


class Solution:
    def rob(self, root: Optional[TreeNode]) -> int:
        def solve(node: Optional[TreeNode]) -> Tuple[int]:
            if not node:
                return (0, 0)
            if not node.left and not node.right:
                return (node.val, 0)

            leftRobbed, leftNotRobbed = solve(node.left)
            rightRobbed, rightNotRobbed = solve(node.right)

            return (
                node.val + leftNotRobbed + rightNotRobbed,
                max(leftRobbed, leftNotRobbed)
                + max(rightRobbed, rightNotRobbed),
            )

        return max(solve(root))

File: medium/0337_house_robber_iii/solution.cpp
#include "../../utils/cpp/nodes.hpp"

class Solution {
public:
  int rob(utils::TreeNode *root) {
    auto solve = [](this auto &&self,
                    utils::TreeNode *node) -> std::pair<int, int> {
      if (!node)
        return {0, 0};
      if (!node->left && !node->right)
        return {node->val, 0};

      auto [leftRobbed, leftNotRobbed] = self(node->left);
      auto [rightRobbed, rightNotRobbed] = self(node->right);

      return {node->val + leftNotRobbed + rightNotRobbed,
              std::max(leftRobbed, leftNotRobbed) +
                  std::max(rightRobbed, rightNotRobbed)};
    };

    auto [robbed, notRobbed] = solve(root);
    return std::max(robbed, notRobbed);
  }
};
